### PR TITLE
[Performance] Binary search for sampling

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
User binary search in `sample_etype_neighbors` to accelerate the process of finding boundary of each edge type.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: src/array/cpu/rowwise_pick.h [hunk 1]
```diff
@@ -310,101 +309,116 @@ COOMatrix CSRRowWisePerEtypePick(
...
-          const IdxType heterogenized_eid =
-              homogenized_eid - eid2etype_offset[heterogenized_etype];
-          et[j] = heterogenized_etype;
-          et_eid[j] = heterogenized_eid;
+        // heterogenized eid is needed when
+        // 1. neighbors of each node is not sorted.
+        // 2. probs_or_mask is not empty as it is indexed by heterogenized eid.
+        if (!rowwise_etype_sorted || has_probs) {
+          et.resize(len);
+          et_eid.resize(len);
```
Comment (BarclayII) [inline src/array/cpu/rowwise_pick.h:321]: What if the condition does not hold (or does it *always* hold)?  Would `et` and `et_eid` be a size-0 vector and cause problems later on?

**Comment by Rhett-Ying** [comment]
@dgl-bot run g4dn.8xlarge bench_sample_neighbors dmlc/master

**Comment by Rhett-Ying** [comment]
@dgl-bot run r6i.16xlarge bench_sample_neighbors dmlc/master

**Comment by Rhett-Ying** [comment]
@peizhou001  I have triggered perf test, please check the results.

File: src/array/cpu/rowwise_pick.h (hunk 1 above)
Comment (peizhou001) [reply src/array/cpu/rowwise_pick.h:321]: It is by design, we use the `IsEmpty` to determine if `etype` and `heterogenized_eid` is calculated in advance in later code.

File: src/array/cpu/rowwise_pick.h [hunk 2]
```diff
@@ -310,101 +309,116 @@ COOMatrix CSRRowWisePerEtypePick(
...
-                    it - eid2etype_offset.begin() - 1;
-                const IdxType heterogenized_eid =
-                    homogenized_eid - eid2etype_offset[heterogenized_etype];
-
-                if (!has_probs ||
-                    IsNullArray(prob_or_mask[heterogenized_etype])) {
-                  // No probability, select all
+                auto element_et = it - eid2etype_offset.begin() - 1;
+                return v < element_et;
+              });
```
Comment (BarclayII) [inline src/array/cpu/rowwise_pick.h:372]: This is nesting two `std::upper_bound`?  I don't quite understand how the outer `std::upper_bound` works.

Comment (peizhou001) [reply src/array/cpu/rowwise_pick.h:372]: The outer `upper_bound` is used to find the end offset of current etype, the inner `upper_bound` is to find the heterogenized etype by searching the `etype_offset` array with `homogeneous eid`.

File: src/array/cpu/rowwise_pick.h (hunk 1 above)
Comment (BarclayII) [reply src/array/cpu/rowwise_pick.h:321]: I saw the `empty()` call in L349-351, but in L384-385, `et` and `et_eid` are still being directly accessed, no?

Comment (peizhou001) [reply src/array/cpu/rowwise_pick.h:321]: In latest code, it will not be accessed in these 2 lines.

**Comment by peizhou001** [comment]
> @peizhou001 I have triggered perf test, please check the results.

Do you mean the `bench_sample_neighbors`? It is for graphbolt sampling.

**Comment by Rhett-Ying** [comment]
FYI. CI is failing due to https://github.com/dmlc/dgl/issues/5902

**Pull request closed without merging**